### PR TITLE
Search: iter through the environment, not libobject

### PR DESCRIPTION
**Pull request by SkySkimmer**
The order of stuff produced by Search is changed, this is OK.

### CODE REVIEW

**Comment by ejgallego** [comment]
Fixes #8495 closes #8228

**Review by ejgallego — APPROVED** [review]
IMHO this is a step in the right direction [as I've argued elsewhere] as we replace a hack with something more principled.

In the medium term tho `Search` should not access the kernel interface [which indeed it is a bit low-level], but better means to maintain a database should be devised.

**Comment by ejgallego** [comment]
> I believe that one could get a much better API than just exposing the kind of "map type" that is used to store things. In particular this PR fights the fact that order cannot be respected, because we lack a timestamp that is otherwise implicit in libstack (here we have 4 maps, libstack is one list). In order to respect the order these map types need to change, or be accompanied by extra data structures: an API could hide that detail and make any related change break less code.
> 
> If we don't want to "bloat" the kernel with this API, I think we should put it elsewhere and expose the `globals` business into an `Internal` sumbmodule (so that it is clear we expose it just to be able to implement a better/safer/robust API on top of it elsewhere).

Indeed I think that environment is not the right API to be consumed by `Search`, as indeed you need quite a bit more information to have efficient indexing, however I do believe that `globals` should be just exposed as a low-level API to be consumed for example by the indexer. IMO not worth having the internal wrapper, just a proper documentation stating that this is a low-level interface so you are on your own.

**Comment by SkySkimmer** [comment]
Here's a commit iterating through safe_env structures instead of the unsafe environment. This allows us to have some order in the iteration. I don't know where to find the dependency order between libraries so they're still unordered, but declarations from one cannot be interleaved with another.

**Comment by SkySkimmer** [comment]
>declarations from one cannot be interleaved with another.

Well they shouldn't be, but it seems they still are for some reason.

**Comment by SkySkimmer** [comment]
>Well they shouldn't be, but it seems they still are for some reason.

Because we stick them through a priority queue, so the order of the initial iteration is irrelevant.

**Comment by ejgallego** [comment]
I am not sure I like this new interface, in fact I don't. The original map was much superior than the list-based one, and should used by the upper layers to build the database.

File: kernel/safe_typing.mli [hunk 1]
```diff
@@ -195,6 +195,19 @@ val export :
...
 
+(** {6 Root extraction } *)
+
+type field =
+  | FModule of ModPath.t
+  | FModuleType of ModPath.t
+  | FConst of Constant.t
+  | FMind of MutInd.t
+
+val roots : safe_environment -> field list
```
Comment (gares) [inline kernel/safe_typing.mli]: Two comments:
- with this approach I don't see the need of making `globals` transparent
- this API should take a filtering predicate, at least for the "module part". For example with `Search ... in Foo Bar` I limit my search to two modules, etc... That is, some filtering can be very cheap to do in this function (module name, label match) , while the more advanced one like "pattern matching" can be left to the caller. See https://coq.inria.fr/distrib/current/refman/proof-engine/ssreflect-proof-language.html#ssr-searching-tool

Comment (ejgallego) [reply kernel/safe_typing.mli]: What you propose shouldn't be in the kernel, much less in safe typing; also the `roots` API is pretty suboptimal when compared the maps, which at least have log access time.

I guess a way to evaluate the API is to answer the question "what kind of efficient indexers can you build _outside_ of the kernel". With `globals`, I can build some for a class of prefix-range based-search queries, with `roots`, none.

Comment (SkySkimmer) [reply kernel/safe_typing.mli]: Here's a change to providing a fold function, so you can do your filtering on the fly.
I don't know about indexing, I'm fine resetting to the first commit though.

Comment (ejgallego) [reply kernel/safe_typing.mli]: > Here's a change to providing a fold function, so you can do your filtering on the fly.
> I don't know about indexing, I'm fine resetting to the first commit though.

Fold is still linear so not too good for me; note that current maps in Coq allow to filter by range and work better for small result sets.

IMVHO from the point of view of `Search` and completion the new API is basically as useless as the old one, so I dunno if that's worth the addition to safe_env.

Comment (SkySkimmer) [reply kernel/safe_typing.mli]: >Fold is still linear so not too good for me; note that current maps in Coq allow to filter by range and work better for small result sets.

The hmaps that we have in the environment have a range filter? How?

>IMVHO from the point of view of Search and completion the new API is basically as useless as the old one, so I dunno if that's worth the addition to safe_env.

True

**Comment by gares** [comment]
I thinkI changed my mind.

The catalog of available lemmas/inductives should be in library, eventually indexed and decorated with metadata such as a timestamp.
And I don't see why this data should be derived from the kernel, it does not even need to be trusted.

What is wrong with the old code?

File: kernel/safe_typing.mli (hunk 1 above)
Comment (ejgallego) [reply kernel/safe_typing.mli]: > The hmaps that we have in the environment have a range filter? How?

Oh indeed, I happened to look at the `MPmap` which is indeed a `CMap`.

**Comment by SkySkimmer** [comment]
>What is wrong with the old code?

I don't like looking at the string tags on objects.

**Comment by gares** [comment]
Then the right fix is to clean up libobject, possibly dropping the idea of a fully extensible but opaque storage and choosing a more rigid one but with a structure that can be traversed.

This may need coordination with @maximedenes

**Review by gares — CHANGES_REQUESTED** [review]
not the right approach imo

**Comment by SkySkimmer** [comment]
I don't understand why we wouldn't look in the kernel for a list of available things.

**Comment by ejgallego** [comment]
> Then the right fix is to clean up libobject, possibly dropping the idea of a fully extensible but opaque storage and choosing a more rigid one but with a structure that can be traversed.

Why should libobject duplicate what's already in the env. If `libobject` is refactored, indeed that should not happen so this PR still would look fairly the same.

**Comment by gares** [comment]
> I don't understand why we wouldn't look in the kernel for a list of available things.

Because the data is already in library, where adding indexing and other features is appropriate.
The fact that the data is in an ugly format is not a reason for looking for it in another place.
You don't want to inspect libobject tags, but you are happy to inspect another data structure that is even more opaque today.

For example, the kernel does not even know the order in which things are declared. You discovered that.
What do we do now? We add that data to the kernel?

I think we should make liboject less hard to inspect.

**Comment by SkySkimmer** [comment]
I still don't like looking at the tags but I don't care enough to argue, 
let's move on.

Gaëtan Gilbert

**Comment by ejgallego** [comment]
> For example, the kernel does not even know the order in which things are declared. You discovered that.
> What do we do now? We add that data to the kernel?
> 
> I think we should make liboject less hard to inspect.

Well, the kernel should know what a module is and that should provide this kind of info.

I don't advocate adding an indexed to the kernel. Clearly, any indexing for search should be performed outside. However, libobject should not contain duplicate information as it does today, and the indexer should work on the data the kernel know.

To that purpose, assume Δ is an update to the kernel env, then
the indexer should also get Δ to keep its database consistent.

In that sense, Gaëtan's indexer is just the identity one; which
admittedly is pretty poor, however that is IMHO better than what
we have today [and allows for more efficient search in some
cases]

**Comment by ejgallego** [comment]
Sad to see this PR go, as it was a net improvement and would have allowed to have nice improvements in Serach for example in the upcoming language server.

Note that the way I see this is:
- kernel maintains the canonical structure:
- libobject maintains Deltas, that are applied to it;
- search listens to deltas and has an efficient engine.

**Pull request closed without merging**